### PR TITLE
migrate realm tests to TS

### PR DESCRIPTION
**Pull request by JacobOscarGunnarsson**

### CODE REVIEW

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 1]
```diff
@@ -0,0 +1,2525 @@
...
+      realm2.close();
+
+      // Open the same in-memory realm again and verify that it is now empty
+      const realm3 = new Realm({ inMemory: true });
+      expect(realm3.schema.length).equals(0);
+
+      // try to open the same realm in persistent mode (should fail as you cannot mix modes)
+      expect(() => new Realm({})).throws("already opened with different inMemory settings.");
+      realm3.close();
+    });
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:648]: Please put spaces between `it`s

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 2]
```diff
@@ -0,0 +1,2525 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      expect(() => realm.write(() => {})).throws("Can't perform transactions on read-only Realms.");
+      realm.close();
+
+      realm = new Realm({ readOnly: true });
+      expect(realm.schema.length).equals(1);
+      expect(realm.isReadOnly).to.be.true;
+      realm.close();
+    });
+  });
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:673]: And spaces between `describe`s

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 3]
```diff
@@ -0,0 +1,2525 @@
...
+  describe("function overwrite", () => {
+    it("bind works on realm", () => {
+      const realm = new Realm({ schema: [] });
+      const oldClose = realm.close.bind(realm);
+      let newCloseCalled = false;
+      realm.close = () => {
+        newCloseCalled = true;
+      };
+      realm.close();
+      expect(newCloseCalled, "The new function should be called").to.be.true;
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:683]: General question, why are we allowing this and testing for it to be doable?

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 4]
```diff
@@ -0,0 +1,2525 @@
...
+
+      expect(realm.isClosed, "The realm should not be closed").to.be.false;
+
+      oldClose();
+
+      expect(realm.isClosed, "The realm should be closed").to.be.true;
+    });
+  });
+
+  describe("object creation", () => {
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:693]: Isn't this covered by the `objects` tests?

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 5]
```diff
@@ -0,0 +1,2525 @@
...
+        });
+      } catch (e: any) {
+        expect(e.stack).not.equals(undefined, "e.stack should not be undefined");
+        expect(e.stack).not.equals(null, "e.stack should not be null");
+        expect(e.stack.indexOf("at failingFunction (") !== -1).to.be.true;
+        expect(e.stack.indexOf("not implemented") !== -1).to.be.true;
+      }
+    });
+  });
+  // TODO wait for app merges
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:921]: Is this your TODO?  Can the TODO be done now?

File: integration-tests/tests/src/tests/sync/realm.ts [hunk 6]
```diff
@@ -0,0 +1,2525 @@
...
+        expectDecimalEqual(obj2.doubleCol, 1.11);
+        expect(obj2.dateCol.getTime()).equals(1);
+        expect(obj2.dataCol.byteLength).equals(1);
+        expect(obj2.objectCol.doubleCol).equals(1);
+        expect(obj2.arrayCol.length).equals(1);
+        expect(obj2.arrayCol[0].doubleCol).equals(1);
+      });
+    });
+
+    it("diffed updates only trigger notificaitons for changed values", async function (this: RealmContext) {
```
Comment (takameyer) [inline integration-tests/tests/src/tests/sync/realm.ts:1192]:
```suggestion
    it("diffed updates only trigger notifications for changed values", async function (this: RealmContext) {
```

Comment (takameyer) [reply integration-tests/tests/src/tests/sync/realm.ts:1192]: And maybe this should already be covered by the notification tests.

**Review by takameyer — APPROVED** [review]
I think there's a lot of refactoring that can be done here.  It seems to me that a lot of the cases are covered elsewhere.  We might want to considering breaking this up into more contextual test cases in different files.  I don't think anyone would be comfortable maintaining this or adding to it.  

That being said, we can merge this for now and take this on later.

File: integration-tests/tests/src/tests/realm-constructor.ts [hunk 7]
```diff
@@ -277,35 +342,74 @@ describe("Realm#constructor", () => {
...
-});
+  describe("in memory construction", () => {
+    it("data is shared among instances", () => {
+      // open in-memory realm instance
+      const realm1 = new Realm({ inMemory: true, schema: [TestObject] });
+      realm1.write(() => {
+        realm1.create("TestObject", [1]);
+      });
+      //@ts-expect-error TYPEBUG: isInMemory property does not exist
+      expect(realm1.isInMemory).to.be.true;
```
Comment (kraenhansen) [inline integration-tests/tests/src/tests/realm-constructor.ts:353]: Very strange that it's been overlooked. You could also just add it here: https://github.com/realm/realm-js/blob/test-migration-master/types/index.d.ts#L1133

**Review by kraenhansen — APPROVED** [review]
Great job migrating these .. there's a lot of code!
You'll become a top contributor soon 🥇

**Pull request merged**